src/server/coprocessor.rs

```diff
@@ -33,14 +92,299 @@ impl SnapshotCoprocessor {
+    let mut e = select::Error::new();
+    e.set_code(1);
+    e.set_msg(format!("{}", err));
+    e
+}
+
+fn encode_key_range(t_info: &TableInfo,
+                    idx_info: &IndexInfo,
+                    ranges: &[KeyRange])
+                    -> Vec<KeyRange> {
+    let t_id = t_info.get_table_id();
+    let idx_id = idx_info.get_index_id();
+    if idx_id == 0 {
+        ranges.iter()
+              .map(|r| {
+                  let mut new_kr = KeyRange::new();
+                  let mut k = table::encode_row_key(t_id, r.get_start());
+                  new_kr.set_start(k);
+                  k = table::encode_row_key(t_id, r.get_end());
+                  new_kr.set_end(k);
+                  new_kr
+              })
+              .collect()
+    } else {
+        ranges.iter()
+              .map(|r| {
+                  let mut new_kr = KeyRange::new();
+                  let mut k = table::encode_index_seek_key(t_id, idx_id, r.get_start());
+                  new_kr.set_start(k);
+                  k = table::encode_index_seek_key(t_id, idx_id, r.get_end());
+                  new_kr.set_end(k);
+                  new_kr
+              })
+              .collect()
+    }
+}
+
+fn get_rows_from_sel(store: &SnapshotStore,
+                     sel: &SelectRequest,
+                     ranges: Vec<KeyRange>)
+                     -> Result<Vec<Row>> {
+    let mut eval = Evaluator::default();
+    let mut rows = vec![];
+    for ran in ranges {
+        let ran_rows = try!(get_rows_from_range(store, sel, ran, &mut eval));
+        rows.extend(ran_rows);
+    }
+    Ok(rows)
+}
+
+fn prefix_next(key: &[u8]) -> Vec<u8> {
+    let mut nk = key.to_vec();
+    if nk.is_empty() {
+        nk.push(0);
+        return nk;
+    }
+    let mut i = nk.len() - 1;
+    loop {
+        if nk[i] == 255 {
+            nk[i] = 0;
+        } else {
+            nk[i] += 1;
+            return nk;
+        }
+        if i == 0 {
+            nk = key.to_vec();
+            nk.push(0);
+            return nk;
+        }
+        i -= 1;
+    }
+}
+
+/// `is_point` checks if the key range represents a point.
+fn is_point(range: &KeyRange) -> bool {
+    range.get_end() == &*prefix_next(range.get_start())
+}
+
+fn get_rows_from_range(store: &SnapshotStore,
+                       sel: &SelectRequest,
+                       mut range: KeyRange,
+                       eval: &mut Evaluator)
+                       -> Result<Vec<Row>> {
+    let mut rows = vec![];
+    if is_point(&range) {
+        if let None = try!(store.get(&Key::from_raw(range.get_start().to_vec()))) {
+            return Ok(rows);
+        }
+        let h = box_try!(table::decode_handle(range.get_start()));
+        if let Some(row) = try!(get_row_by_handle(store, sel, h, eval)) {
+            rows.push(row);
+        }
+    } else {
+        let mut seek_key = range.take_start();
+        loop {
+            let mut res = try!(store.scan(Key::from_raw(seek_key), 1));
```

Comment:
No, actually scan is implemented by seek, and we always use a different key to seek next time.